File: src/distribution/dis.py3/FVA.py
from scipy.optimize import linprog
import numpy as np
import sys

# FVA in python
def calculate_flux_variabilty(dataDictionary):
    # Get some stuff from the DF -
    STM = dataDictionary["stoichiometric_matrix"]
    OBJVECTOR = dataDictionary["objective_coefficient_array"]
    number_of_fluxes = np.size(OBJVECTOR)
    # Get Flux bounds sequence from the DF -
    FluxBounds = dataDictionary["default_flux_bounds_array"]
    # Get species bounds
    SpeciesBounds = dataDictionary["species_bounds_array"]
    # Get numbers
    NUM_Unbalanced = dataDictionary["extra_species_num"]
    NUM_Speices = len(SpeciesBounds)
    NUM_Balanced = NUM_Speices - NUM_Unbalanced

    # Equality constraints
    Aeq = STM[NUM_Unbalanced:NUM_Speices]
    # Formulate the bV -
    bVEq = np.zeros(NUM_Balanced)

    # Inequality constraints
    UNBALANCED_STM = STM[0:NUM_Unbalanced]
    bVLB = [-SpeciesBounds[i][0] for i in range(NUM_Unbalanced)]
    bVUB = [SpeciesBounds[i][1] for i in range(NUM_Unbalanced)]
    A = np.concatenate((UNBALANCED_STM, -1*UNBALANCED_STM))
    bV = np.concatenate((bVUB, bVLB))

    # initialize the output array -
    calculated_flux_array = np.zeros((np.size(OBJVECTOR), 3))

    # Phase 1: solve the original problem as a minimization problem-
    # default parameter setting of linprog: maxiter=1000, disp=False, tol=1.0E-12,
    opt = {"disp": True, "tol": 1.0E-8}
    FBA_result = linprog(OBJVECTOR, A_ub=A, b_ub=bV, A_eq=Aeq, b_eq=bVEq,
                 bounds=FluxBounds, options=opt)
    if FBA_result.status == 1:
        print("\nrerun linprog with 10*maxiter")
        opt = {"disp": True, "tol": 1.0E-8, "maxiter": 10000}
        FBA_result = linprog(OBJVECTOR, A_ub=A, b_ub=bV, A_eq=Aeq, b_eq=bVEq,
                     bounds=FluxBounds, options=opt)
    if FBA_result.status == 2:
        sys.exit()
    for i in range(number_of_fluxes):
        calculated_flux_array[i, 0] = FBA_result.x[i]

    # Phase 2: solve the FVA problem
    new_A = np.concatenate((A, [OBJVECTOR]))
    new_bV = np.concatenate((bV, [FBA_result.fun]))
    new_OBJVECTOR = np.zeros(number_of_fluxes)
    # main FVA loop -
    for round_index in range(2):
        # min or max
        if round_index == 0:
            OptDir = 1
        else:
            OptDir = -1
        for i in range(number_of_fluxes):
            new_OBJVECTOR[i] = OptDir
            opt = {"disp": True, "tol": 1.0E-8}
            res = linprog(new_OBJVECTOR, A_ub=new_A, b_ub=new_bV, A_eq=Aeq,
                          b_eq=bVEq, bounds=FluxBounds, options=opt)
            if res.status == 1:
                opt = {"disp": True, "tol": 1.0E-8, "maxiter": 10000}
                res = linprog(new_OBJVECTOR, A_ub=new_A, b_ub=new_bV, A_eq=Aeq,
                              b_eq=bVEq, bounds=FluxBounds, options=opt)
            if res.status != 2:
                calculated_flux_array[i, round_index+1] = OptDir*res.fun
            else:
                print("ERROR: Unable to solve modified problem for flux indexed as", i)
            new_OBJVECTOR[i] = 0

    return calculated_flux_array

File: src/distribution/dis.m/FVA.m
% FVA calling GLPK
% Solving the initial problem first, then does FVA
function [Flux_Array] = FVA(dataDictionary, MIN_MAX_FLAG)

	% Get some stuff from the DD -
	STM = dataDictionary('stoichiometric_matrix');
	[NUM_Species, NUM_Var] = size(STM);
	OBJVECTOR = dataDictionary('objective_coefficient_array');

	% Get Flux bounds from the DD -
	FluxBounds = dataDictionary('default_flux_bounds_array');
	FluxLB = FluxBounds(:,1);
	FluxUB = FluxBounds(:,2);
    % Formulate the VARTYPE vector (Flux)-
    VARTYPE = blanks(NUM_Var)';
    for id = 1:NUM_Var
        VARTYPE(id, 1) = 'C';
    end

	% Get species bounds
	SpeciesBounds = dataDictionary('species_bounds_array');
	% Get numbers
	NUM_Unbalanced = dataDictionary('extra_species_num');
	NUM_Balanced = NUM_Species - NUM_Unbalanced;

	% Equality constraints
	Aeq = STM((NUM_Unbalanced+1):NUM_Species, :);
	% Formulate the bV -
	bVEq = zeros(NUM_Balanced,1);

	% Inequality constraints
	UNBALANCED_STM = STM(1:NUM_Unbalanced, :);
	bVLB = SpeciesBounds(1:NUM_Unbalanced, 1);
	bVUB = SpeciesBounds(1:NUM_Unbalanced, 2);

    % Formulate the matrix of constraint coefficients
	A = [UNBALANCED_STM ; UNBALANCED_STM ; Aeq];
    % Formulate the column array containing the right-hand side value for
    % each constraint in the constraint matrix.
	bV = [bVUB ; bVLB; bVEq];
    % Formulate the CTYPE vector for each constraint
    CTYPE = blanks(length(bV))';
    for id = 1:NUM_Unbalanced
        CTYPE(id, 1) = 'U';
        CTYPE(id+NUM_Unbalanced, 1) = 'L';
    end
    offset = 2*NUM_Unbalanced;
    for id = (1+offset):(NUM_Balanced+offset)
        CTYPE(id) = 'S';
    end

    % ---------------------------------------------
    Flux_Array = zeros(NUM_Var, 3);

    % Phase One: Solve the initial problem P first -
	% Set the sense flag
    if (MIN_MAX_FLAG == -1)
		sense = -1;
    else
	 	sense = 1;
    end
	% Set some values for the optional parameters
	param.msglev = 3;
    param.tmlim = 3600;
	% Call the glpk solver -
	[FLOW,FVAL,P_status, P_extra] = glpk(OBJVECTOR, A, bV, FluxLB, FluxUB, CTYPE, VARTYPE, sense, param);
    if (P_status == 109)
        fprintf('failed to solve in 1h, try 10h...');
        param.tmlim = 3600*10;
        [FLOW,FVAL,P_status, P_extra] = glpk(OBJVECTOR, A, bV, FluxLB, FluxUB, CTYPE, VARTYPE, sense, param);
    end
    if (P_status == 109)
        fprintf('failed to find the optima after running for 10h');
    end
    for i = 1:NUM_Var
        Flux_Array(i, 1) = FLOW(i);
    end

    % Phase Two: FVA
    param.presol = 0;  % Turn off presolver
    new_OBJ = zeros(NUM_Var, 1);
    new_A = [A; OBJVECTOR];
    new_bV = [bV; FVAL];
    if (MIN_MAX_FLAG == -1)
        new_CT = [CTYPE; 'L'];
    else
        new_CT = [CTYPE; 'U'];
    end
    for round_id = 1:2
        if round_id == 1
            new_sense = 1;
        else
            new_sense = -1;
        end
        for i = 1:NUM_Var
            new_OBJ(i) = 1;
            param.tmlim = 3600;
            [F,fopt,S, E] = glpk(new_OBJ, new_A, new_bV, FluxLB, FluxUB, new_CT, VARTYPE, new_sense, param);
            if (S == 109)
                fprintf('failed to solve in 1h, try 10h...');
                param.tmlim = 3600*10;
                [F,fopt,S, E] = glpk(new_OBJ, new_A, new_bV, FluxLB, FluxUB, new_CT, VARTYPE, new_sense, param);
            end
            if (S == 109)
                fprintf('failed to find the optima after running for 10h');
            end
            Flux_Array(i, round_id + 1) = fopt;
            new_OBJ(i) = 0;
        end
    end

return;
